### PR TITLE
Fix enum mapping validation

### LINKED ISSUE #11119: BackedEnum type with interface validation mismatch
### BC Break Report

|    Q        |   A
|------------ | ------
| BC Break    | yes
| Version     | 2.17.1

#### Summary

After upgrading to version 2.17.1 I've encountered this validation error:

```
$ bin/console doctrine:schema:validate

[FAIL] The entity-class App\Entity\CertificateEvent mapping is invalid:
 * The field 'App\Entity\CertificateEvent#type' has the property type 'App\Enum\EventType' that differs from the metadata field type 'string' returned by the 'string' DBAL type.
```

#### Previous behavior
Schema validation pass

#### Current behavior
Schema validation fails

#### How to reproduce

```php
interface EventType
{
}

enum CertificateEventType: string implements EventType
{
}
```

the mapping is:
```xml
<entity name="App\Entity\CertificateEvent">
    <field name="type" type="string" column="type" enum-type="App\Enum\CertificateEventType" nullable="false" />
```

CertificateEvent looks like:
```php
class CertificateEvent extends CaseEvent
{
    public function __construct(
        CertificateEventType $type
    ) {
      parent::__construct($case, $type);
    }
```

```php
abstract class CaseEvent
{
    public function __construct(
        protected EventType $type,
    ) {
```

### PR DESCRIPTION
**Pull request by yceruto**
Closes #11119

### CODE REVIEW

File: lib/Doctrine/ORM/Tools/SchemaValidator.php [hunk 1]
```diff
@@ -406,6 +407,25 @@ function (array $fieldMapping) use ($class): ?string {
                             );
                         }
 
+                        if (
+                            isset($fieldMapping['enumType'])
+                            && $propertyType !== $fieldMapping['enumType']
+                            && interface_exists($propertyType)
+                            && is_a($fieldMapping['enumType'], $propertyType, true)
+                        ) {
+                            if ($metadataFieldType === (string) (new ReflectionEnum($fieldMapping['enumType']))->getBackingType()) {
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Tools/SchemaValidator.php]: You're basing this on a comparison between `$metadataFieldType` and the backing type of the enum, yet only one of those appears in the error message.

File: lib/Doctrine/ORM/Tools/SchemaValidator.php [hunk 2]
```diff
@@ -406,6 +407,25 @@ function (array $fieldMapping) use ($class): ?string {
...
+                                return null;
+                            }
+
+                            return sprintf(
+                                "The field '%s#%s' has the metadata enumType '%s' that differs from the metadata field type '%s'.",
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Tools/SchemaValidator.php]: So let's mention it here please?

```suggestion
                                "The field '%s#%s' has the metadata enumType '%s' with a backing type of %s that differs from the metadata field type '%s'.",
```

File: lib/Doctrine/ORM/Tools/SchemaValidator.php [hunk 3]
```diff
@@ -406,6 +407,28 @@ function (array $fieldMapping) use ($class): ?string {
...
+                            }
+
+                            return sprintf(
+                                "The field '%s#%s' has the metadata enumType '%s' with a backing type of '%s' that differs from the metadata field type '%s'.",
+                                $class->name,
+                                $fieldName,
+                                $fieldMapping['enumType'],
+                                $backingType,
+                                $metadataFieldType
+                            );
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Tools/SchemaValidator.php:439]: Should this be extracted somehow, and reused on line 400? There could be the same improvement here (right?). Or maybe both conditions should somehow be combined? I having a hard time figuring out why you are not touching code already dealing with enums within this class :thinking:

File: lib/Doctrine/ORM/Tools/SchemaValidator.php [hunk 4]
```diff
@@ -406,6 +407,28 @@ function (array $fieldMapping) use ($class): ?string {
                             );
                         }
 
+                        if (
+                            isset($fieldMapping['enumType'])
+                            && $propertyType !== $fieldMapping['enumType']
+                            && interface_exists($propertyType)
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Tools/SchemaValidator.php:423]: Why do we need to check that it's an interface?

File: lib/Doctrine/ORM/Tools/SchemaValidator.php (hunk 3 above)
Comment (yceruto) [reply lib/Doctrine/ORM/Tools/SchemaValidator.php:439]: I'm having a hard time figuring out the same concern. I'm still checking, but they seem like different scenarios.

File: lib/Doctrine/ORM/Tools/SchemaValidator.php (hunk 4 above)
Comment (yceruto) [reply lib/Doctrine/ORM/Tools/SchemaValidator.php:423]: For two reasons: 1) The static analyzer complained about passing a non-class-string as the second argument of the `is_a` function; 2) It's only possible for enums to implement interfaces.

File: lib/Doctrine/ORM/Tools/SchemaValidator.php (hunk 3 above)
Comment (yceruto) [reply lib/Doctrine/ORM/Tools/SchemaValidator.php:439]: Note that in this use case, the property type is an interface, so we don't really know what it's for. The only thing we can check is whether the mapped enum type is an implementation of that interface, and whether it matches the mapping type, nothing more.

Anyway, I will give it a try to see if by inverting the first block, I can merge something there.

**Review by greg0ire — DISMISSED** [review]
After re-reading the first case, my brain hurts but I think it's indeed really different. I think you actually went with the right way, combining the conditions is probably going to produce something very hard to understand.

**Comment by yceruto** [comment]
Still, I improved the first error message to include the backing type information.

File: tests/Doctrine/Tests/ORM/Functional/Ticket/GH11037/GH11037Test.php [hunk 5]
```diff
@@ -40,8 +40,9 @@ public function testMetadataFieldTypeNotCoherentWithEntityPropertyType(): void
 
         self::assertEquals(
             [
-                "The field 'Doctrine\Tests\ORM\Functional\Ticket\GH11037\InvalidEntityWithTypedEnum#status1' has the property type 'Doctrine\Tests\ORM\Functional\Ticket\GH11037\StringEntityStatus' that differs from the metadata field type 'int' returned by the 'integer' DBAL type.",
+                "The field 'Doctrine\Tests\ORM\Functional\Ticket\GH11037\InvalidEntityWithTypedEnum#status1' has the property type 'Doctrine\Tests\ORM\Functional\Ticket\GH11037\StringEntityStatus' with a backing type of 'string' that differs from the metadata field type 'int'.",
```
Comment (greg0ire) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/GH11037/GH11037Test.php:43]: Ah yes, that's way better 👍

**Review by SenseException — APPROVED** [review]

**Comment by greg0ire** [comment]
Thanks @yceruto !

**Pull request merged**